security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient.go

```diff
@@ -57,67 +53,86 @@ type federatedTokenResponse struct {
 
 // TokenExchanger for google securetoken api interaction.
 type Plugin struct {
-	hTTPClient *http.Client
+	httpClient  *http.Client
+	credFetcher security.CredFetcher
+	backoff     time.Duration
 }
 
 // NewPlugin returns an instance of secure token service client plugin
-func NewPlugin() security.TokenExchanger {
-	caCertPool, err := x509.SystemCertPool()
-	if err != nil {
-		stsClientLog.Errorf("Failed to get SystemCertPool: %v", err)
-		return nil
-	}
-	return Plugin{
-		hTTPClient: &http.Client{
-			Timeout: httpTimeOutInSec * time.Second,
-			Transport: &http.Transport{
-				TLSClientConfig: &tls.Config{
-					RootCAs: caCertPool,
-				},
-			},
+func NewPlugin(credFetcher security.CredFetcher) *Plugin {
+	stsClientLog.SetOutputLevel(log.DebugLevel)
+	return &Plugin{
+		httpClient: &http.Client{
+			Timeout: httpTimeout,
 		},
+		backoff:     time.Millisecond * 50,
+		credFetcher: credFetcher,
 	}
 }
 
-// ExchangeToken exchange oauth access token from trusted domain and k8s sa jwt.
-func (p Plugin) ExchangeToken(ctx context.Context, credFetcher security.CredFetcher, trustDomain, k8sSAjwt string) (
-	string /*access token*/, time.Time /*expireTime*/, int /*httpRespCode*/, error) {
-	aud := constructAudience(credFetcher, trustDomain)
-	var jsonStr = constructFederatedTokenRequest(aud, k8sSAjwt)
-	req, _ := http.NewRequest("POST", SecureTokenEndpoint, bytes.NewBuffer(jsonStr))
-	req.Header.Set("Content-Type", contentType)
-
-	resp, err := p.hTTPClient.Do(req)
-	errMsg := "failed to call token exchange service. "
-	if err != nil || resp == nil {
-		statusCode := http.StatusServiceUnavailable
-		// If resp is not null, return the actually status code returned from the token service.
-		// If resp is null, return a service unavailable status and try again.
-		if resp != nil {
-			statusCode = resp.StatusCode
-			errMsg += fmt.Sprintf("HTTP status: %s. Error: %v", resp.Status, err)
-		} else {
-			errMsg += fmt.Sprintf("HTTP response empty. Error: %v", err)
+func retryable(code int) bool {
+	return code >= 500 && !(code == 501 || code == 505 || code == 511)
+}
+
+func (p *Plugin) requestWithRetry(reqBytes []byte) ([]byte, error) {
+	attempts := 0
+	var lastError error
+	for attempts < 5 {
+		attempts++
+		req, err := http.NewRequest("POST", SecureTokenEndpoint, bytes.NewBuffer(reqBytes))
```

Comment:
Shall we move the req generation above the for loop?

Reply:
This is actually required, an http request cannot be re-used

Reply:
Cool. Thanks.